Spring Security Bonus Module

brclark opened the pull request:
Contains curriculum for Java Authentication Bonus Module. This module covers a multi-step process to use Spring Security framework and design some Coding Events features that would make use of multiple user roles and privileges.

The lessons are being added to `Chapter 19 Authentication/Next Steps/Bonus Module`.

The sequence of lessons are:

1. **User Data**: Add relationships to `User` and resources like `Event` to allow users to own their data
2. **Add Services**: Make use of Spring `@Service` classes to handle logic between `@Controller` and `@Repository`, which also makes more use of DTO objects for data translation
3. **Add User Roles & Privileges**: Create models for roles and privileges that can be associated with users
4. **Spring Security Features**: Fully Implement Authentication and Authorization in Spring Security 6, to handle request filters and session management
5. **Role Based Access**: Limit access to Coding Events based on the assigned User role, such as base `ROLE_USER`, event organizer `ROLE_ORGANIZER`, and admin `ROLE_ADMIN`.

The associated code is in the following project:
https://github.com/brclark/CodingEventsJava/tree/user-data

brclark committed:
Add new section to Authentication on User Owned Data

Added new curriculum text for User Owned Data to the Authentication
chapter. First draft so it needs to be edited.

This section describes how users can own the data they create in
Coding Events.

TODO:
 - Create tutorial video
 - Edit this section

brclark committed:
Add section for DTOs & Services to Authentication

- Add content for DTOs and Services
- TODO:
    - Add video content
    - Complete rest of code description

brclark committed:
Add text description for EventService and EventCategoryService

- Added fixes for UserService description
- Add descriptions for EventService and EventCategoryService
- Add description to refactor controllers to use services

brclark committed:
Add text walkthrough for User Roles & Privileges

Add text based content for how to add roles and privileges to Coding Events
and associate them to users. This content does not include any changes to
functionality.

brclark committed:
Move PasswordEncoder to a new configuration and refactor User

The PasswordEncoder is no longer stored inside the User model as
a static instance. It has been moved to a configuration class and
is a managed bean of the Spring framework.

brclark committed:
Add missing methods in UserService description

Description was missing explanation of `save` and `validateUser` methods.
Added explanations and code snippets.

brclark committed:
Add InitialDataLoader description to User Roles & Privileges

Added text description for how to preload data in to the database
for user roles and privileges.

brclark committed:
Move User Roles & Privileges walkthrough to Authentication Next Steps

Create a bonus module within "Next Steps" to house the curriculum for
user roles and privileges and spring security. This is within the
Authentication chapter

brclark committed:
Add SecurityService to User Roles & Privileges

Add description for initial implementation of `SecurityService` to
User Roles & Privileges curriculum.

brclark committed:
Add Spring Security 6 instruction set for Authentication Bonus

Adds content for Spring Security bonus module:
- Instructions on using spring security framework for user authentication
- Instructions for spring security authorization framework using
security context filters

brclark commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Coding Events Java - User Roles/Spring Security](https://trello.com/c/73jPNsTq/1691-coding-events-java-user-roles-spring-security)

brclark committed:
Add Role Based Access instructions

Create user attendance relationship and UI CRUD
Add `PreAuthorize` annotation to controllers to restrict access
based on role
Update navigation to reflect the current user's role

brclark committed:
Fixes part 1 - going throuh the User Owned Data section

Fix mermaid issues and clean up grammar/spelling

brclark commented on content/authentication/reading/add-service-dto/_index.md:
```diff
@@ -301,4 +354,358 @@ Add this method to your `UserService` below the fields.
+
+Similar to our `EventService`, we must add methods to the `EventCategoryService`
+to expose functionality of the `EventCategoryRepository` to the controllers.
+
```
Don't include code for `EventCategoryService`. Just include instructions.

Also consider adding interface for CategoryService and EventService

brclark commented on content/authentication/reading/add-service-dto/_index.md:
```diff
@@ -301,4 +354,358 @@ Add this method to your `UserService` below the fields.
+how the sequence diagram above is organized.
+
+#### `EventCategoryController`
+
```
Get rid of the code in this section and move it after `EventController`

brclark commented on content/authentication/reading/add-service-dto/_index.md:
````diff
@@ -301,4 +354,358 @@ Add this method to your `UserService` below the fields.
+    }
+```
+
+```java {hl_lines="3"}
````
`hl_lines="4"`

brclark commented on content/authentication/reading/add-service-dto/_index.md:
````diff
@@ -301,4 +354,358 @@ Add this method to your `UserService` below the fields.
+    }
+```
+
+```java {wrap="true" hl_lines="2 10"}
````
Mention the removal of `HttpSession` and service save method

brclark commented on content/authentication/reading/add-service-dto/_index.md:
```diff
@@ -301,4 +354,358 @@ Add this method to your `UserService` below the fields.
+        EventDetails details = new EventDetails(eventDTO.getDescription(), eventDTO.getContactEmail());
+        event.setEventDetails(details);
+
+        event.setEventCategory(categoryRepository.findById(eventDTO.getCategoryId()).get());
```
use .orElse(null) here, also consider using categoryService instead of Repository

brclark commented on content/authentication/reading/add-service-dto/_index.md:
````diff
@@ -301,4 +354,358 @@ Add this method to your `UserService` below the fields.
+        return "events/index";
+    }
+```
+
````
Add instructions between code snippets

brclark commented on content/authentication/reading/add-service-dto/_index.md:
```diff
@@ -143,6 +143,57 @@ public class EventCategoryDTO {
+Modify the `User` constructor so that it takes in `String pwHash` as an argument
+and uses it to set the field directly, removing the call to `encode`.
+
+Lastly, remove the `isMatchingPassword` method and replace it with a getter for
```
Need to add fix/section for `AuthenticationController` refactor  which was using `isMatchingPassword` as well

Should include instructions for adding `UserService` to `AuthenticationController`

brclark reviewed (commented):
Review of User Owned Data, and Add Service DTOs instructions

brclark committed:
Edit the add-service-dto section with some updates from review

Clean up some of the code based on review and errors
Change the description to force the reader to implement more of
`EventCategory` on their own.
Add instructions for `AuthenticationController` updates to use
`UserService`

brclark committed:
Fixes from user-roles-privileges proofread

Some wording updates

brclark committed:
Spring Security fixes from proofread

Update the mermaid chart for class diagram
Fix small mistakes in wording or missing variables/method names

gildedgardenia commented on content/authentication/next-steps/_index.md:
```diff
@@ -16,6 +16,9 @@ lastMod: # UPDATE ANY TIME CHANGES ARE MADE
 1. View [this guide](https://spring.io/guides/gs/securing-web/) that walks through the creation of a spring MVC web application using Spring Security.
-1. [Baeldung Spring Security Basic Authentication](https://www.baeldung.com/spring-security-basic-authentication) walkthrough. This will will walk through the process of basic authentication with Spring with an MVC application.
\ No newline at end of file
+1. [Baeldung Spring Security Basic Authentication](https://www.baeldung.com/spring-security-basic-authentication) walkthrough. This will will walk through the process of basic authentication with Spring with an MVC application.
```
```suggestion
1. [Baeldung Spring Security Basic Authentication](https://www.baeldung.com/spring-security-basic-authentication) walkthrough. This will walk through the process of basic authentication with Spring with an MVC application.
```

gildedgardenia commented on content/authentication/next-steps/bonus-module/user-data/_index.md:
```diff
@@ -0,0 +1,287 @@
+to log in before they can access any features of the site. However, users
+will want the data that they create to be associated with their account. How
+can we make sure that the events and categories that a user create will be
+associated with their account?
```
```suggestion
will want the data that they create to be associated with their account. How
can we ensure that the events and categories a user creates are
associated with their account?
```

gildedgardenia commented on content/authentication/next-steps/bonus-module/user-data/_index.md:
```diff
@@ -0,0 +1,287 @@
+can we make sure that the events and categories that a user create will be
+associated with their account?
+
+Let's associate the new data with the logged-in user using one-to-many database relationships.
```
```suggestion
Let's use a one-to-many relationship to associate the new data with the logged-in user.
```

brclark replied:
I rewrote the intro to this section to reduce the redundancy of explanation

gildedgardenia commented on content/authentication/next-steps/bonus-module/user-data/_index.md:
```diff
@@ -0,0 +1,287 @@
+Users *own* their data when the entities that they create (events, categories, etc) have their
+`user_id` associated with each new entity as a foreign key. That would allow us to, 
+say "get all events for a specific user". We need to set up a **One-To-Many** relationship
+between the `User` model and the different data models (`Event`, `EventCategory`, `Tag`).
```
```suggestion
To set up a one-to-many relationship between the `User` model and the other data models (`Event`, `EventCategory`, and `Tag`), we need to add `user_id` to the other data models as a foreign key. 
```

gildedgardenia replied:
You already said that we need to set up a one-to-many relationship above so the final sentence of this paragraph is redundant.

gildedgardenia commented on content/authentication/next-steps/bonus-module/user-data/_index.md:
```diff
@@ -0,0 +1,287 @@
+
+Open the `Event` model and add a `User` field to the field definitions. We will call it `creator`
+as it's a reference to the user that created the event, and we will give it a `@ManyToOne` annotation
+as there can be many events associated with a single creator.
```
```suggestion
Open the `Event` model and add a `User` field to the field definitions. Call it `creator` and give it a `@ManyToOne` annotation.
```

gildedgardenia commented on content/authentication/next-steps/bonus-module/user-data/_index.md:
````diff
@@ -0,0 +1,287 @@
+```
+
+We'll also need to add a getter and setter. Add the following to your getters
+and setters in `Event` class.
````
```suggestion
We'll also need to add a getter and setter for `creator`.
```

gildedgardenia commented on content/authentication/next-steps/bonus-module/user-data/_index.md:
````diff
@@ -0,0 +1,287 @@
+```
+
+`Event` is not the only model that we want to be user-owned. **Repeat the above
+steps** to add the `creator` field/getters/setters to `EventCategory`.
````
```suggestion
Repeat the above steps to add the `creator` field/getters/setters to `EventCategory`.
```

gildedgardenia commented on content/authentication/next-steps/bonus-module/user-data/_index.md:
```diff
@@ -0,0 +1,287 @@
+
+### Saving the `User` when creating new data
+
+We can store an associated User as the creator of an Event/Category/Tag. Next
```
```suggestion
We can store an associated `User` as the creator of an `Event`/`EventCategory`/`Tag`. Next
```

gildedgardenia commented on content/authentication/next-steps/bonus-module/user-data/_index.md:
```diff
@@ -0,0 +1,287 @@
+### Saving the `User` when creating new data
+
+We can store an associated User as the creator of an Event/Category/Tag. Next
+we need to make sure that the currently logged-in User is set as the creator
```
```suggestion
we need to make sure that the currently logged-in user is set as the creator
```

gildedgardenia commented on content/authentication/next-steps/bonus-module/user-data/_index.md:
```diff
@@ -0,0 +1,287 @@
+We can store an associated User as the creator of an Event/Category/Tag. Next
+we need to make sure that the currently logged-in User is set as the creator
+before saving new entries. Let's update the `EventController` first to set
+the `User creator` field in new events.
```
```suggestion
the `creator` field in new events.
```

gildedgardenia commented on content/authentication/next-steps/bonus-module/user-data/_index.md:
```diff
@@ -0,0 +1,287 @@
+            model.addAttribute("title", "Create Event");
+            model.addAttribute("categories", eventCategoryRepository.findAllByCreator(currUser));
+            return "events/create";
+        }
```
Is this method missing a closing bracket?

gildedgardenia commented on content/authentication/next-steps/bonus-module/user-data/_index.md:
````diff
@@ -0,0 +1,287 @@
+In the `displayEventDetails` method, let's add:
+
+```java{hl_lines="2-5"}
+    @GetMapping("detail")
+    public String displayEventDetails(@RequestParam Integer eventId, Model model, HttpSession session) {
+        User currUser = authController.getUserFromSession(session);
+
+        Optional<Event> result = eventRepository.findByIdAndCreator(eventId, currUser);
````
If you are only showing part of a method, I would add some contextual clues so they know for sure where to add the code. See the suggestion below for how I would do it.

gildedgardenia replied:
```suggestion
    @GetMapping("detail")
    public String displayEventDetails(@RequestParam Integer eventId, Model model, HttpSession session) {
        User currUser = authController.getUserFromSession(session);

        Optional<Event> result = eventRepository.findByIdAndCreator(eventId, currUser);
        
        // Rest of our method
        }
```

gildedgardenia commented on content/authentication/next-steps/bonus-module/role-based-access/_index.md:
```diff
@@ -0,0 +1,584 @@
+
+## Adding Role-Based Features for Users - TEXT
+
+The first portion of this lesson will be add event attendance relationships
```
```suggestion
The first portion of this lesson will add event attendance relationships
```

gildedgardenia commented on content/authentication/next-steps/bonus-module/role-based-access/_index.md:
```diff
@@ -0,0 +1,584 @@
+attendance for an event by the current user.
+
+In `EventController`, let's first add a route at `GET /events/attending` that
+will load the table of events that the current user has rsvped to:
```
```suggestion
will load the table of events that the current user has RSVP'd to:
```

gildedgardenia commented on content/authentication/next-steps/bonus-module/role-based-access/_index.md:
````diff
@@ -0,0 +1,584 @@
+    private RoleRepository roleRepository;
+```
+
+#### Update `getAuthorities` to pull roles & privileges for User
````
```suggestion
#### Update `getAuthorities` to pull roles & privileges for `User`
```

gildedgardenia commented on content/authentication/next-steps/bonus-module/role-based-access/_index.md:
````diff
@@ -0,0 +1,584 @@
+```
+
+Now someone can successfully register as a different user type with
+`ROLE_ORGANIZER` set.
````
```suggestion
Now when someone checks the option for "I am an event organizer", their user type for `ROLE ORGANIZER` is set.
```

gildedgardenia commented on content/authentication/next-steps/bonus-module/role-based-access/_index.md:
```diff
@@ -0,0 +1,584 @@
+similar to the `th:if` attribute that will selectively include an HTML element
+depending on the condition.
+
+We will either use:
```
```suggestion
We will use one of the following:
```

gildedgardenia commented on content/authentication/next-steps/bonus-module/spring-security/_index.md:
```diff
@@ -0,0 +1,731 @@
+
+### Authentication in Spring Security 6
+
+First, it is important to take a look at the [Spring Security Authentication
```
```suggestion
Before diving in, take a look at the [Spring Security Authentication
```

gildedgardenia commented on content/authentication/next-steps/bonus-module/spring-security/_index.md:
```diff
@@ -0,0 +1,731 @@
+`IUserService` should include two method definitions, which mirror the
+methods we have already implemented in `UserService`.
+
+* `User findByUsername(String username);`
+* `User save(RegisterFormDTO registration);`
```
I would encourage you to turn some of the unordered lists into ordered lists. Ordered lists are easier for screenreader users to navigate.

gildedgardenia commented on content/authentication/next-steps/bonus-module/spring-security/_index.md:
```diff
@@ -0,0 +1,731 @@
+1. `@EnableTransactionManagement` annotation allows our Spring project to use
+the `@Transactional` annotation elsewhere
+
+To set up the `SecurityFilterChain` we will use method chaining to build an
```
```suggestion
To set up the `SecurityFilterChain`, we will use method chaining to build an
```

gildedgardenia commented on content/authentication/next-steps/bonus-module/spring-security/_index.md:
```diff
@@ -0,0 +1,731 @@
+
+#### Update `EventController`
+
+Our goal for `EventController` will be the same as `EventCategoryController`
+--- we want to remove all references to `AuthenticationController` and the
```
```suggestion
Our goal for `EventController` will be the same as `EventCategoryController`: to remove all references to `AuthenticationController` and the
```

gildedgardenia commented on content/authentication/next-steps/bonus-module/spring-security/_index.md:
```diff
@@ -0,0 +1,731 @@
+
+First, **remove** the `AuthenticationController` autowired field.
+
+Next, we need one new method in `EventCategoryService` for retrieving a
+category by an id but also assuming the current user. We want to prevent
```
```suggestion
Next, we need one new method in `EventCategoryService` for retrieving a
category that the current user created by the category id. We want to prevent
```

gildedgardenia reviewed (commented):
Here are my suggestions for you! Thanks for your work on this, Ben!

brclark committed:
Add fixes from peer review

Clean up language and grammar in the lessons.
Implement some fixes to code snippets.

brclark committed:
Updates from testing and tutorial making

Add numbered lists instead of unordered lists.
Remove unneeded sections based on code refactor

brclark committed:
Clean up intro to bonus module, Tag instructions, and code fixes

Add intro description to the Bonus Module
Leave hints about implementing the `Tag` class changes in the instructions
Update code to match latest version of project

brclark committed:
Remove mentions of video lessons from spring security bonus module

This leaves the bonus module as a strictly text-based walkthrough

brclark commented on content/authentication/next-steps/bonus-module/_index.md:
```diff
@@ -0,0 +1,25 @@
+and roles & privileges to restrict access to the Coding Events based
+on assigned user roles.
+
+### TODO: Add Embedded Intro video
```
I think including the preview video that I made might be useful here, to show students what they will be building by the end of the walkthrough
